File: Python/No_Repeats_Please.py
#coding:utf-8

def hasNotRepeat(string):
    if len(string) <=1:
        return True
    else:
        for char in set(string):
            index = string.find(char)
            if index!=-1 and index!=len(string)-1:
                if string[index]==string[index+1]:
                    return False
        return True

def findPermutations(string):
    if len(string) < 2:
        return string
    
    permutationsArray = [] 

    for index,char in enumerate(string):
        remainingChars = string[0:index] + string[index+1:]

        for permutation in findPermutations(remainingChars):
            if hasNotRepeat(char+permutation):
                permutationsArray.append(char+permutation)

    return permutationsArray


def permAlone(string):
    return len(findPermutations(string))

if __name__ == '__main__':

    print(permAlone("aab"))
    

File: JavaScript/No_Repeats_Please.js
function hasNotRepeat(str){
    if(str.length<=1)
        return true;
    else{
        let mySet = new Set(str);
        for(let character of mySet){
            let index = str.indexOf(character);
            if(index!=str.length-1){
                if(str[index]==str[index+1])
                    return false;
            }
        }
        return true;
    }
}

let findPermutations = (string) => {
    if (!string || typeof string !== "string"){
      return "Please enter a string"
    } else if (string.length < 2 ){
      return string
    }
  
    let permutationsArray = [] 
     
    for (let i = 0; i < string.length; i++){
      let char = string[i]
  
      //if (string.indexOf(char) != i)
      //continue
  
      let remainingChars = string.slice(0, i) + string.slice(i + 1, string.length)
  
      for (let permutation of findPermutations(remainingChars)){
          if(hasNotRepeat(char + permutation)){
            permutationsArray.push(char + permutation)
          }
         }
    }
    return permutationsArray;
  }


  function permAlone(str) {
    return findPermutations(str).length;
  }
  
  console.log(permAlone('aab'));
  